### PR TITLE
Add eval mode to Mask R-CNN TPU runner

### PR DESCRIPTION
**Pull request by sprt**
This introduces changes that allow one to compute the mAP scores for Mask R-CNN on TPUs using `python train_tpu.py --mode=eval`.

### CODE REVIEW

File: references/detection/train_tpu.py [hunk 1]
```diff
@@ -65,7 +78,15 @@ def get_transform(train):
...
+        '--mode', type=str, default='test',
+        choices=('eval', 'test', 'train'),
+        help='eval computes the AP scores, test runs inference on a few sample images, '
+        'train trains the model')
+    parser.add_argument(
+        '--num-eval', type=int, default=50,
+        help='Number of validation set images to use in eval mode')
+    parser.add_argument(
+        '--num_cores', type=int, default=8,
         help='Number of TPU cores to train on (8 for single v2/v3 Cloud TPU)')
```
Comment (taylanbil) [inline references/detection/train_tpu.py:90]: does this work on v2s?

File: references/detection/train_tpu.py [hunk 2]
```diff
@@ -98,7 +119,7 @@ def random_colour_masks(mask):
 
 
 def overlay_output(img, masks, boxes, pred_cls, alpha=0.5):
-    for i in range(len(masks)):
+    for i in range(len(boxes)):
```
Comment (taylanbil) [inline references/detection/train_tpu.py]: is `boxes` of different length sometimes? is this intended?

File: torchvision/models/detection/transform.py [hunk 3]
```diff
@@ -56,6 +62,17 @@ def normalize(self, image):
         return (image - mean[:, None, None]) / std[:, None, None]
 
     def resize(self, image, target):
+        if self.is_xla:
+            # Resize to a fixed size on TPU to avoid dynamicity.
+            new_h, new_w = 600, 800
```
Comment (taylanbil) [inline torchvision/models/detection/transform.py:67]: should these be command line arguments, or the first size the class sees, instead of fixed values?

File: torchvision/models/detection/transform.py [hunk 4]
```diff
@@ -56,6 +62,17 @@ def normalize(self, image):
...
+            image = torch.nn.functional.interpolate(
+                image[None], size=(new_h, new_w), mode='bilinear', align_corners=False)[0]
+
+            if target is None:
+                return image, target
+ 
+            raise NotImplementedError
```
Comment (taylanbil) [inline torchvision/models/detection/transform.py:74]: can you explain why we only return if target is None here?

File: references/detection/engine.py [hunk 5]
```diff
@@ -80,13 +82,20 @@ def evaluate(model, data_loader, device):
...
+        if device.type == 'cuda':
+            torch.cuda.synchronize()
 
-        torch.cuda.synchronize()
         model_time = time.time()
         outputs = model(image)
 
+        if device.type == 'xla':
+            for output in outputs:
+                torch_xla._XLAC._xla_sync_multi(list(output.values()), devices=[])
```
Comment (taylanbil) [inline references/detection/engine.py]: does it work if we use `xm.mark_step` here? If it does not change much, better to use the api instead of the underscore functions.

Comment (jysohn23) [reply references/detection/engine.py]: Or would the `xm.mark_step` be *after* line 99?

File: references/detection/train_tpu.py [hunk 6]
```diff
@@ -128,28 +149,23 @@ def process_prediction(output, threshold=0.05):
...
 
-    if not use_cpu:
-        # Send to single TPU device
-        device = xm.xla_device()
-        torch_xla._XLAC._xla_set_default_device(str(device))
-        model = model.to(device)
-        image_tensor = image_tensor.to(device)
+    with _bench('inference'):
+        output = model([image_tensor])
+        torch_xla._XLAC._xla_sync_multi(list(output[0].values()), devices=[])
```
Comment (jysohn23) [inline references/detection/train_tpu.py]: This may be a better play to try out if `xm.mark_step` works instead of that C binding as Taylan suggested.

File: references/detection/train_tpu.py [hunk 7]
```diff
@@ -128,28 +149,23 @@ def process_prediction(output, threshold=0.05):
...
+
+    with _bench('cpu postprocessing'):
+        masks, pred_boxes, pred_class = process_prediction(output)
+        image = overlay_output(image, masks, pred_boxes, pred_class)
+        image.save(os.path.join('/coco/out', os.path.basename(image_path)))
 
-    masks, pred_boxes, pred_class = process_prediction(output)
-    image = overlay_output(image, masks, pred_boxes, pred_class)
-    image.save(os.path.join('/coco/out', os.path.basename(image_path)))
+    print()
```
Comment (jysohn23) [inline references/detection/train_tpu.py]: Intended `print()`?

File: references/detection/train_tpu.py [hunk 8]
```diff
@@ -229,14 +245,42 @@ def do_train(num_cores):
...
 
-
-
 def main():
     FLAGS = parse_flags()
     logger.info("FLAGS: {}".format(FLAGS))
 
-    do_prediction('/coco/train2017/000000566234.jpg', use_cpu=FLAGS.use_cpu)
-    #do_train(FLAGS.num_cores)
+    def mark_step():
```
Comment (jysohn23) [inline references/detection/train_tpu.py:250]: Just a suggestion, but do you think using this `mark_step()` may yield good info in eval?

File: torchvision/models/detection/transform.py [hunk 9]
```diff
@@ -56,6 +62,17 @@ def normalize(self, image):
         return (image - mean[:, None, None]) / std[:, None, None]
 
     def resize(self, image, target):
+        if self.is_xla:
+            # Resize to a fixed size on TPU to avoid dynamicity.
+            new_h, new_w = 600, 800
+            image = torch.nn.functional.interpolate(
+                image[None], size=(new_h, new_w), mode='bilinear', align_corners=False)[0]
```
Comment (jysohn23) [inline torchvision/models/detection/transform.py:69]: Mind explaining what `image[None]` is doing here?

File: references/detection/train_tpu.py (hunk 1 above)
Comment (sprt) [reply references/detection/train_tpu.py:90]: The `--num_cores` flag is from @jysohn23's work back in June so I would assume so?

File: references/detection/train_tpu.py (hunk 2 above)
Comment (sprt) [reply references/detection/train_tpu.py]: No, I had made that change when debugging the masks as I thought it might be that `len(boxes) != len(masks)`, but it's never the case (rightfully so) and I forgot to revert this line back to the original. I can revert it in a new commit if you want, but the output will be the same 😃

File: references/detection/train_tpu.py (hunk 1 above)
Comment (taylanbil) [reply references/detection/train_tpu.py:90]: presumably w/ small backbones it may work, although it may be valuable to verify that first unless @jysohn23 already verified?

File: references/detection/train_tpu.py (hunk 2 above)
Comment (taylanbil) [reply references/detection/train_tpu.py]: yeah you could push a new commit to this branch and it'll update the pr.

File: references/detection/train_tpu.py (hunk 1 above)
Comment (jysohn23) [reply references/detection/train_tpu.py:90]: IIRC, it should work if batch sizes are small enough.

File: references/detection/engine.py (hunk 5 above)
Comment (taylanbil) [reply references/detection/engine.py]: yeah that would be better. less barriers = more goodness

File: references/detection/train_tpu.py [hunk 10]
```diff
@@ -128,28 +149,23 @@ def process_prediction(output, threshold=0.05):
...
+    with _bench('cpu postprocessing'):
+        masks, pred_boxes, pred_class = process_prediction(output)
+        image = overlay_output(image, masks, pred_boxes, pred_class)
+        image.save(os.path.join('/coco/out', os.path.basename(image_path)))
 
-    masks, pred_boxes, pred_class = process_prediction(output)
-    image = overlay_output(image, masks, pred_boxes, pred_class)
-    image.save(os.path.join('/coco/out', os.path.basename(image_path)))
+    print()
+    print(torch_xla._XLAC._xla_metrics_report())
```
Comment (taylanbil) [inline references/detection/train_tpu.py]: when we run this multi-core or pods, this would print a lot of things to the screen. Instead, let's use `xm.master_print`

File: references/detection/train_tpu.py [hunk 11]
```diff
@@ -229,14 +245,42 @@ def do_train(num_cores):
...
+        dataset_val, _ = get_dataset('coco', 'val', T.ToTensor())
+        dataset_val = torch.utils.data.Subset(dataset_val, range(FLAGS.num_eval))
+
+        data_loader_val = torch.utils.data.DataLoader(
+            dataset_val, batch_size=1, shuffle=False, num_workers=0,
+            collate_fn=utils.collate_fn)
+
+        evaluate(model, data_loader_val, device=device, move_tensors_to_device=False)
+    elif FLAGS.mode == 'test':
+        mark_step()
```
Comment (taylanbil) [inline references/detection/train_tpu.py]: this is probably not needed, no operations have been performed at this point iiuc.

File: torchvision/models/detection/rpn.py [hunk 12]
```diff
@@ -73,7 +73,7 @@ def generate_anchors(scales, aspect_ratios, dtype=torch.float32, device="cpu"):
         hs = (h_ratios[:, None] * scales[None, :]).view(-1)
 
         base_anchors = torch.stack([-ws, -hs, ws, hs], dim=1) / 2
-        return base_anchors.round()
+        return base_anchors
```
Comment (taylanbil) [inline torchvision/models/detection/rpn.py:76]: for my own curiosity; is this commented because it's a `_local_scalar_dense` call?

File: torchvision/models/detection/transform.py (hunk 4 above)
Comment (sprt) [reply torchvision/models/detection/transform.py:74]: If you scroll down you can see it's the original behavior of that method. Basically, in inference mode (as opposed to training), `target` is None and therefore there's no operations to run on it. I haven't implemented resizing of the boxes/masks/keypoints for training mode yet, so that's why I throw an exception.

File: references/detection/train_tpu.py (hunk 7 above)
Comment (sprt) [reply references/detection/train_tpu.py]: Yup, just so that the `# of boxes: ...` isn't stuck to the metrics report :smiley:

File: references/detection/train_tpu.py (hunk 8 above)
Comment (sprt) [reply references/detection/train_tpu.py:250]: Hmm, the `evaluate()` function already calls `xm.mark_step()`. Or do you mean benchmarking it might be useful?

File: torchvision/models/detection/transform.py (hunk 9 above)
Comment (sprt) [reply torchvision/models/detection/transform.py:69]: TBH not sure, but it's the original behavior of the method if you scroll down further to line 89

File: references/detection/train_tpu.py (hunk 7 above)
Comment (sprt) [reply references/detection/train_tpu.py]: Removed it

File: torchvision/models/detection/rpn.py (hunk 12 above)
Comment (sprt) [reply torchvision/models/detection/rpn.py:76]: `aten::round` more specifically

File: torchvision/models/detection/transform.py (hunk 3 above)
Comment (sprt) [reply torchvision/models/detection/transform.py:67]: Think making those configurable is definitely a better idea yeah, although might have to be done in another PR since I'm not sure I'll be able to have time to do this

File: references/detection/train_tpu.py (hunk 8 above)
Comment (jysohn23) [reply references/detection/train_tpu.py:250]: Yeah the benchmarking version of `mark_step`. But an optional 😄

**Review by jysohn23 — APPROVED** [review]

**Comment by taylanbil** [comment]
Thanks for all the explanation!

**Review by taylanbil — APPROVED** [review]
